File: src/main/java/my/study/repository/design/pattern/builder/java/ComplexObject.java
package my.study.repository.design.pattern.builder.java;

/**
 * @author Djh on 2018/7/27 15:25
 * E-Mail ：1544579459@qq.com
 */
public class ComplexObject {

    private int mAge;
    private String mSex;
    private String mName;
    private String mFatherName;

    private ComplexObject(Builder builder) {
        mSex = builder.mSex;
        mAge = builder.mAge;
        mName = builder.mName;
        mFatherName = builder.mFatherName;
    }

    public static class Builder {
        private int mAge;
        private String mSex;
        private String mName;
        private String mFatherName;

        public Builder setName(String name) {
            mName = name;
            return this;
        }

        public Builder setAge(int age) {
            mAge = age;
            return this;
        }

        public Builder setSex(String sex) {
            mSex = sex;
            return this;
        }

        public Builder setFatherName(String fatherName) {
            mFatherName = fatherName;
            return this;
        }

        public ComplexObject build() {
            return new ComplexObject(this);
        }
    }
}

File: src/main/java/my/study/repository/design/pattern/builder/kotlin/ComplexObject.kt
package my.study.repository.design.pattern.builder.kotlin

/**
 * @author djh on  2018/10/3 20:15
 * @E-Mail 1544579459@qq.com
 */
class ComplexObject private constructor(builder: Builder) {

    val mAge: Int
    private var mSex: String
    private var mName: String
    private var mFatherName: String

    init {
        mAge = builder.mAge
        mSex = builder.mSex
        mName = builder.mName
        mFatherName = builder.mFatherName
    }


    class Builder {

        var mAge = 0
        var mSex = ""
        var mName = ""
        var mFatherName = ""

        fun setAge(age: Int): Builder {
            mAge = age
            return this
        }

        fun setSex(sex: String): Builder {
            mSex = sex
            return this
        }

        fun setName(name: String): Builder {
            mName = name
            return this
        }

        fun setFatherName(fatherName: String): Builder {
            mFatherName = fatherName
            return this
        }

        fun build(): ComplexObject {
            return ComplexObject(this)
        }

    }
}